File: src/main/java/me/itzisonn_/meazy/runtime/environment/FunctionDeclarationEnvironment.java
package me.itzisonn_.meazy.runtime.environment;

import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.parser.ast.expression.ParameterExpression;
import me.itzisonn_.meazy.runtime.value.FunctionValue;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

import java.lang.constant.ClassDesc;
import java.util.List;
import java.util.Optional;
import java.util.Set;

/**
 * Adds to Environment ability to declare functions
 */
@NullMarked
public interface FunctionDeclarationEnvironment extends Environment {
    /**
     * Declares given function in this environment
     * TODO
     */
    FunctionValue declareFunction(String id, List<ParameterExpression> parameters, @Nullable DataType returnDataType, FunctionEnvironment functionEnvironment);

    /**
     * @param id Id
     * @param parameters Parameters
     * @return Declared function with given id and args or null
     */
    default Optional<FunctionValue> getFunction(String id, List<ClassDesc> parameters) {
        main:
        for (FunctionValue functionValue : getFunctions()) {
            if (!functionValue.getId().equals(id)) continue;

            List<ParameterExpression> functionParameters = functionValue.getParameters();
            if (functionParameters.size() != parameters.size()) continue;

            for (int i = 0; i < parameters.size(); i++) {
                ClassDesc functionParameterClassDesc = functionParameters.get(i).getDataType().getClassDesc();
                ClassDesc parameterClassDesc = parameters.get(i);
                if (!EnvironmentUtils.isInstanceOf(this, parameterClassDesc, functionParameterClassDesc)) continue main;
            }

            return Optional.of(functionValue);
        }

        return Optional.empty();
    }

    /**
     * @return All declared functions
     */
    Set<FunctionValue> getFunctions();
}

File: src/main/java/me/itzisonn_/meazy/runtime/environment/FunctionDeclarationEnvironment.kt
package me.itzisonn_.meazy.runtime.environment;

import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.parser.ast.expression.ParameterExpression;
import org.jspecify.annotations.NullMarked;

import java.util.List;
import java.util.Optional;
import java.util.Set;

/**
 * Adds to Environment ability to declare functions
 */
@NullMarked
public interface FunctionDeclarationEnvironment extends Environment {
    /**
     * Declares given function in this environment
     * TODO
     */
    void declareFunction(FunctionEnvironment functionEnvironment);

    /**
     * @param id Id
     * @param args Parameters
     * @return Declared function with given id and args or null
     */
    default Optional<FunctionEnvironment> getFunction(String id, List<DataType> args) {
        main:
        for (FunctionEnvironment functionEnvironment : getFunctions()) {
            if (!functionEnvironment.getId().equals(id)) continue;

            List<ParameterExpression> parameters = functionEnvironment.getParameters();
            if (parameters.size() != args.size()) continue;

            for (int i = 0; i < args.size(); i++) {
                DataType parameter = parameters.get(i).getDataType();
                DataType arg = args.get(i);
                if (!DataType.matches(this, arg, parameter)) continue main;
            }

            return Optional.of(functionEnvironment);
        }

        return Optional.empty();
    }

    /**
     * @return All declared functions
     */
    Set<FunctionEnvironment> getFunctions();
}